# **Data Cleaning**

## Objectives

* Prepare the data sets for for further analysis

## Inputs

* /workspace/housing-sales-price-prediction-dashboard/inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv
* /workspace/housing-sales-price-prediction-dashboard/inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/inherited_houses.csv

## Outputs

* ./outputs/datasets/clean_house_prices_records.csv
* ./outputs/datasets/clean_inherited_houses.csv

---

# Change working directory

* We are assuming you will store the notebooks in a sub folder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/workspace/housing-sales-price-prediction-dashboard/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/workspace/housing-sales-price-prediction-dashboard'

---

# Import Packages

In [4]:
import numpy as np
import pandas as pd

---

## Load Data

In [5]:
df = pd.read_csv(f"inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv")
print(df.shape)

(1460, 24)


In [6]:
df_inherited = pd.read_csv(f"inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/inherited_houses.csv")
print(df_inherited.shape)

(4, 23)


---

# Dataset Analysis

Note that there is no `id` field to mark row data uniqueness.

There are however differences in datatypes between the house_prices_records and inherited_houses.
* Certain fields are of type int64 in house_prices_records and of type float64 in inherited_houses and visa versa.

## Individual variable analysis

Inspect `WoodDeckSF` variable

In [7]:
df_wooddecksf = df.loc[df['WoodDeckSF'].notnull()]
df_wooddecksf

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0          856     854.0           3.0           No         706          GLQ   
13        1494       0.0           3.0           Av           0          Unf   
20        1158    1218.0           4.0           Av           0          Unf   
23        1060       0.0           3.0           No         840          GLQ   
34        1561       0.0           2.0           No        1153          GLQ   
...        ...       ...           ...          ...         ...          ...   
1419      1844       0.0           3.0           Av        1110          Rec   
1422       848       0.0           1.0           Av         686          GLQ   
1431       958       0.0           2.0           No         958          LwQ   
1455       953     694.0           3.0           No           0          Unf   
1459      1256       0.0           3.0           No         830          BLQ   

      BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotFrontage  \
0           150            0.0         548          RFn  ...         65.0   
13         1494            NaN         840          RFn  ...         91.0   
20         1158            NaN         853          RFn  ...        101.0   
23          200            NaN         572          NaN  ...         44.0   
34          408            NaN         556          Fin  ...         60.0   
...         ...            ...         ...          ...  ...          ...   
1419        734          216.0         540          RFn  ...          NaN   
1422        162            NaN         420          Fin  ...         37.0   
1431          0            NaN         440          RFn  ...          NaN   
1455        953            NaN         460          RFn  ...         62.0   
1459        136            0.0         276          Fin  ...         75.0   

      MasVnrArea OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  \
0          196.0          61            5            7          856   
13         306.0          33            5            7         1494   
20         380.0         154            5            8         1158   
23           0.0         110            7            5         1040   
34         246.0          47            5            9         1561   
...          ...         ...          ...          ...          ...   
1419       312.0          73            5            6         1844   
1422       170.0           0            5            6          848   
1431         0.0          60            6            6          958   
1455         0.0          40            5            6          953   
1459         0.0          68            6            5         1256   

      WoodDeckSF  YearBuilt  YearRemodAdd  SalePrice  
0            0.0       2003          2003     208500  
13         160.0       2006          2007     279500  
20         240.0       2005          2006     325300  
23         100.0       1976          1976     129900  
34         203.0       2005          2005     277500  
...          ...        ...           ...        ...  
1419         0.0       1969          1969     223000  
1422       140.0       2003          2003     136500  
1431         0.0       1976          1976     143750  
1455         0.0       1999          2000     175000  
1459       736.0       1965          1965     147500  

[155 rows x 24 columns]

In [8]:
df_wooddecksf['WoodDeckSF'].value_counts().sort_index(ascending=False)


736.0     1
550.0     1
466.0     1
431.0     1
416.0     1
382.0     1
364.0     1
356.0     1
355.0     1
351.0     1
349.0     1
344.0     1
328.0     1
307.0     1
300.0     2
296.0     1
292.0     1
289.0     1
288.0     2
252.0     1
240.0     2
216.0     2
208.0     1
206.0     1
203.0     1
197.0     1
196.0     2
192.0     5
188.0     1
185.0     1
180.0     1
168.0     2
164.0     1
162.0     1
160.0     2
147.0     1
146.0     1
145.0     1
144.0     3
143.0     1
141.0     1
140.0     2
139.0     1
127.0     1
120.0     4
117.0     1
116.0     1
106.0     1
105.0     1
104.0     1
100.0     4
84.0      1
78.0      1
44.0      1
36.0      1
33.0      1
24.0      1
0.0      78
Name: WoodDeckSF, dtype: int64

---

### Variables to consider dropping

#### EnclosedPorch - Enclosed porch area in square feet
This variable contains more than 90% null values in the house price dataset. In the inherited dataset the value for this variable is 0 for all 4 properties, meaning the porch is not enclosed. Based on this variable will add no value to the sale price analysis.

#### WoodDeckSF - Wood deck area in square feet
This variable contains 89% null values in the house price dataset. In the inherited dataset this field contains valid values, however, due to the lack of comparative data this variable may add no value at all. Furthermore, doing a value count shows the data contains diverse sizes and not enough uniqueness, and little to no exact matches to the inherited house dataset

In [9]:
df = df.drop(['EnclosedPorch', 'WoodDeckSF'], axis=1)
df.head()

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       856     854.0           3.0           No         706          GLQ   
1      1262       0.0           3.0           Gd         978          ALQ   
2       920     866.0           3.0           Mn         486          GLQ   
3       961       NaN           NaN           No         216          ALQ   
4      1145       NaN           4.0           Av         655          GLQ   

   BsmtUnfSF  GarageArea GarageFinish  GarageYrBlt  ...  LotArea LotFrontage  \
0        150         548          RFn       2003.0  ...     8450        65.0   
1        284         460          RFn       1976.0  ...     9600        80.0   
2        434         608          RFn       2001.0  ...    11250        68.0   
3        540         642          Unf       1998.0  ...     9550        60.0   
4        490         836          RFn       2000.0  ...    14260        84.0   

   MasVnrArea  OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  YearBuilt  \
0       196.0           61            5            7          856       2003   
1         0.0            0            8            6         1262       1976   
2       162.0           42            5            7          920       2001   
3         0.0           35            5            7          756       1915   
4       350.0           84            5            8         1145       2000   

   YearRemodAdd  SalePrice  
0          2003     208500  
1          1976     181500  
2          2002     223500  
3          1970     140000  
4          2000     250000  

[5 rows x 22 columns]

In [10]:
df_inherited = df_inherited.drop(['EnclosedPorch', 'WoodDeckSF'], axis=1)
df_inherited.head()

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       896         0             2           No       468.0          Rec   
1      1329         0             3           No       923.0          ALQ   
2       928       701             3           No       791.0          GLQ   
3       926       678             3           No       602.0          GLQ   

   BsmtUnfSF  GarageArea GarageFinish  GarageYrBlt  ...  KitchenQual LotArea  \
0      270.0       730.0          Unf       1961.0  ...           TA   11622   
1      406.0       312.0          Unf       1958.0  ...           Gd   14267   
2      137.0       482.0          Fin       1997.0  ...           TA   13830   
3      324.0       470.0          Fin       1998.0  ...           Gd    9978   

   LotFrontage  MasVnrArea  OpenPorchSF  OverallCond  OverallQual  \
0         80.0         0.0            0            6            5   
1         81.0       108.0           36            6            6   
2         74.0         0.0           34            5            5   
3         78.0        20.0           36            6            6   

   TotalBsmtSF  YearBuilt  YearRemodAdd  
0        882.0       1961          1961  
1       1329.0       1958          1958  
2        928.0       1997          1998  
3        926.0       1998          1998  

[4 rows x 21 columns]

---

### Get null value count on the house price dataset

In [11]:
df.isna().sum()

1stFlrSF          0
2ndFlrSF         86
BedroomAbvGr     99
BsmtExposure      0
BsmtFinSF1        0
BsmtFinType1    114
BsmtUnfSF         0
GarageArea        0
GarageFinish    162
GarageYrBlt      81
GrLivArea         0
KitchenQual       0
LotArea           0
LotFrontage     259
MasVnrArea        8
OpenPorchSF       0
OverallCond       0
OverallQual       0
TotalBsmtSF       0
YearBuilt         0
YearRemodAdd      0
SalePrice         0
dtype: int64

---

### Variables to consider transforming or imputing

#### 2ndFlrSF - Second floor square feet

* Inspect `2ndFlrSF` variable

In [12]:
df['2ndFlrSF'].value_counts().sort_index()

0.0       781
110.0       1
167.0       1
192.0       1
208.0       1
         ... 
1611.0      1
1796.0      1
1818.0      1
1872.0      1
2065.0      1
Name: 2ndFlrSF, Length: 401, dtype: int64

* `86` variables of `1460` contain null values. It appears that if there is no second floor the value would be set to `0`. More than 50% of values for this variable are 0. Therefore one may deduce imputing the null values with 0 would add value.
* We prepare the pipeline to use `ArbitraryNumberImputer` to impute `0` into the null variables

---

#### BedroomAbvGr - Bedrooms above grade (does NOT include basement bedrooms)

* Inspect `BedroomAbvGr` variable

In [13]:
df['BedroomAbvGr'].value_counts().sort_index()

0.0      6
1.0     46
2.0    333
3.0    749
4.0    199
5.0     20
6.0      7
8.0      1
Name: BedroomAbvGr, dtype: int64

* `99` variables of `1460` contain null values. A value count shows that only 6 records have a `0` for this variable. All 4 inherited properties contain values above zero. Imputing the null values with 0 may have no effect on the sales price analysis.
* We prepare the pipeline to use `ArbitraryNumberImputer` to impute `0` into the null variables

---

In [14]:
from feature_engine.imputation import ArbitraryNumberImputer
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
      ( '2ndFlrSF',  ArbitraryNumberImputer(arbitrary_number=0,
                                                variables=['2ndFlrSF', 'BedroomAbvGr']) )
])
pipeline

Pipeline(steps=[('2ndFlrSF',
                 ArbitraryNumberImputer(arbitrary_number=0,
                                        variables=['2ndFlrSF',
                                                   'BedroomAbvGr']))])

In [15]:
# pipeline.fit(df)
# df = pipeline.transform(df)
df = pipeline.fit_transform(df)

In [16]:
df.isna().sum()

1stFlrSF          0
2ndFlrSF          0
BedroomAbvGr      0
BsmtExposure      0
BsmtFinSF1        0
BsmtFinType1    114
BsmtUnfSF         0
GarageArea        0
GarageFinish    162
GarageYrBlt      81
GrLivArea         0
KitchenQual       0
LotArea           0
LotFrontage     259
MasVnrArea        8
OpenPorchSF       0
OverallCond       0
OverallQual       0
TotalBsmtSF       0
YearBuilt         0
YearRemodAdd      0
SalePrice         0
dtype: int64

---

#### BsmtFinType1 - Rating of basement finished area

* Inspect `BsmtFinType1` variable

In [17]:
df['BsmtFinType1'].value_counts().sort_index()

ALQ     202
BLQ     136
GLQ     385
LwQ      70
None     31
Rec     126
Unf     396
Name: BsmtFinType1, dtype: int64

* Inspect `BsmtExposure` variable

In [18]:
df['BsmtExposure'].value_counts().sort_index()

Av      221
Gd      134
Mn      114
No      953
None     38
Name: BsmtExposure, dtype: int64

In [19]:
df[df['BsmtFinType1'].isna()].query('BsmtExposure=="None"').sort_values(by=['BsmtExposure'])

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
90        1040       0.0           2.0         None           0          NaN   
182       1340       0.0           3.0         None           0          NaN   
362        495    1427.0           4.0         None           0          NaN   
736       1040       0.0           2.0         None           0          NaN   
1049       930       0.0           2.0         None           0          NaN   
1179      1124       0.0           3.0         None           0          NaN   

      BsmtUnfSF  GarageArea GarageFinish  GarageYrBlt  ...  LotArea  \
90            0         420          Unf       1950.0  ...     7200   
182           0         252          NaN       1957.0  ...     9060   
362           0         672          RFn       2003.0  ...     7301   
736           0         400          Unf       1949.0  ...     8544   
1049          0         308          Unf       1946.0  ...    11100   
1179          0           0         None          NaN  ...     8335   

     LotFrontage  MasVnrArea  OpenPorchSF  OverallCond  OverallQual  \
90          60.0         0.0           29            5            4   
182         60.0        98.0            0            6            5   
362         64.0       500.0            0            5            7   
736         60.0         0.0            0            4            3   
1049        60.0         0.0            0            7            4   
1179        77.0         0.0           36            5            5   

      TotalBsmtSF  YearBuilt  YearRemodAdd  SalePrice  
90              0       1950          1950     109900  
182             0       1957          2006     120000  
362             0       2003          2003     198500  
736             0       1950          1950      93500  
1049            0       1946          2006      84900  
1179            0       1954          1954      93000  

[6 rows x 22 columns]

* `114` variables of `1460` contain null values. There are only `31` properties with no basement.
* `BsmtExposure` however contains no null variables and on comparing the two fields I established that there are `6` rows are set to `None`, meaning that they have no basement, where `BsmtFinType1` is null. For these 6 rows the `BsmtFinType1` variable may be imputed with `None`.

In [20]:
query_condition = (df.BsmtExposure == 'None') & (df['BsmtFinType1'].isnull())
df['BsmtFinType1'] = np.where(query_condition, 'None', df['BsmtFinType1'])

In [21]:
df[df['BsmtFinType1'].isna()].query('BsmtExposure=="None"').sort_values(by=['BsmtExposure'])

Empty DataFrame
Columns: [1stFlrSF, 2ndFlrSF, BedroomAbvGr, BsmtExposure, BsmtFinSF1, BsmtFinType1, BsmtUnfSF, GarageArea, GarageFinish, GarageYrBlt, GrLivArea, KitchenQual, LotArea, LotFrontage, MasVnrArea, OpenPorchSF, OverallCond, OverallQual, TotalBsmtSF, YearBuilt, YearRemodAdd, SalePrice]
Index: []

[0 rows x 22 columns]

In [22]:
df['BsmtFinType1'].isna().sum()

108

---

* Inspect `BsmtFinSF1` variable. Type 1 finished square feet

In [23]:
df[df['BsmtFinType1'].isna()].query('BsmtFinSF1==0').sort_values(by=['BsmtFinSF1'])

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
22        1795       0.0           3.0           No           0          NaN   
1408       741     622.0           3.0           No           0          NaN   
1379       754     640.0           3.0           No           0          NaN   
1375      1571       0.0           3.0           Gd           0          NaN   
1342      1284     885.0           3.0           No           0          NaN   
1337       693       0.0           2.0           No           0          NaN   
1318      1787       0.0           3.0           Gd           0          NaN   
1219       672     546.0           3.0           No           0          NaN   
1203      1630       0.0           3.0           No           0          NaN   
1202       884     464.0           3.0           No           0          NaN   
1082      1459       0.0           3.0           No           0          NaN   
1051      1316       0.0           3.0           Av           0          NaN   
965        783     701.0           3.0           No           0          NaN   
890        672     252.0           0.0           Mn           0          NaN   
866       1646       0.0           3.0           Av           0          NaN   
805       1428       0.0           3.0           No           0          NaN   
742       1349       0.0           3.0           No           0          NaN   
716       1576     978.0           3.0           No           0          NaN   
101        977     755.0           3.0           No           0          NaN   
124       1586       0.0           3.0           No           0          NaN   
157        774    1194.0           4.0           No           0          NaN   
243        626     591.0           3.0           No           0          NaN   
248        840     828.0           3.0           No           0          NaN   
374        847    1081.0           4.0           No           0          NaN   
1423      1575     626.0           4.0           Av           0          NaN   
394        735     299.0           2.0           No           0          NaN   
431       1088     110.0           4.0           No           0          NaN   
573        774     656.0           3.0           No           0          NaN   
581       2042       0.0           3.0           No           0          NaN   
584        989     584.0           3.0           No           0          NaN   
601        780     595.0           3.0           No           0          NaN   
653        756     713.0           3.0           No           0          NaN   
422        988       0.0           2.0           No           0          NaN   
1435      1537       0.0           3.0           No           0          NaN   

      BsmtUnfSF  GarageArea GarageFinish  GarageYrBlt  ...  LotArea  \
22         1777         534          RFn       2002.0  ...     9742   
1408        622         528          Unf       1966.0  ...     7740   
1379        384         400          Fin       2007.0  ...     9735   
1375       1571         722          Fin       2007.0  ...    10991   
1342       1284         647          RFn       2002.0  ...     9375   
1337        693           0         None          NaN  ...     4118   
1318       1753         748          RFn       2001.0  ...    14781   
1219        672           0         None          NaN  ...     1680   
1203       1630         451          Unf       2000.0  ...     9750   
1202        884         216          Unf       1960.0  ...     6000   
1082       1459         527          RFn       2002.0  ...     8749   
1051       1316         440          Fin       2007.0  ...    11175   
965         783         393          Fin       2005.0  ...    10237   
890         672         576          Unf       2003.0  ...     8064   
866        1638         870          RFn       2007.0  ...    10656   
805        1417         554          RFn       2008.0  ...    12274  

* Next we look at `BsmtFinSF1` which contains no null variables. We search for `BsmtFinType1` with nulls and `BsmtFinSF1` with value `0`, meaning `0` finished square feet, which means unfinished. We find there are 34 records. Therefore, we can impute these records with `Unf` which means unfinished for `BsmtFinType1` variable.

In [24]:
query_condition = (df.BsmtFinSF1 == 0) & (df['BsmtFinType1'].isnull())
df['BsmtFinType1'] = np.where(query_condition, 'None', df['BsmtFinType1'])

In [25]:
df[df['BsmtFinType1'].isna()].query('BsmtFinSF1==0').sort_values(by=['BsmtFinSF1'])

Empty DataFrame
Columns: [1stFlrSF, 2ndFlrSF, BedroomAbvGr, BsmtExposure, BsmtFinSF1, BsmtFinType1, BsmtUnfSF, GarageArea, GarageFinish, GarageYrBlt, GrLivArea, KitchenQual, LotArea, LotFrontage, MasVnrArea, OpenPorchSF, OverallCond, OverallQual, TotalBsmtSF, YearBuilt, YearRemodAdd, SalePrice]
Index: []

[0 rows x 22 columns]

In [26]:
df['BsmtFinType1'].isna().sum()

74

---

* Inspect `GarageFinish` variable. Interior finish of the garage

In [27]:
df['GarageFinish'].isna().sum()

162

* There are `162` null variables for `GarageFinish`

In [28]:
df['GarageFinish'].value_counts().sort_index()

Fin     313
None     73
RFn     366
Unf     546
Name: GarageFinish, dtype: int64

In [29]:
df.loc[df.GarageFinish=="None",'GarageArea'].value_counts()


0    73
Name: GarageArea, dtype: int64

* We found above, where `GarageFinish=="None"`, meaning there is no garage, we also find `GarageArea` to be `0`.

In [30]:
df.loc[df.GarageFinish.isnull(),'GarageArea'].value_counts()

0      8
440    7
240    4
308    4
484    3
      ..
866    1
529    1
696    1
678    1
840    1
Name: GarageArea, Length: 124, dtype: int64

* Therefore, where `GarageFinish` is null we can check if `GarageArea` is `0` and if so we can impute `None` on `GarageFinish`.
* based on the above query, only 8 rows will be affected.
* For the remaining roecords we will assume that the garages are unfinished and hence impute `Unf` on `GarageFinish`.

In [31]:
query_condition = (df.GarageArea == 0) & (df['GarageFinish'].isnull())
df['GarageFinish'] = np.where(query_condition, 'None', df['GarageFinish'])

In [32]:
df['GarageFinish'].isna().sum()

154

In [33]:
from feature_engine.imputation import CategoricalImputer
pipeline = Pipeline([
      ( 'categorical_imputer', CategoricalImputer(imputation_method='missing',
                                                  fill_value='Unf',
                                                  variables=['GarageFinish']) )
])
pipeline

Pipeline(steps=[('categorical_imputer',
                 CategoricalImputer(fill_value='Unf',
                                    variables=['GarageFinish']))])

In [34]:
df = pipeline.fit_transform(df)

In [35]:
df['GarageFinish'].isna().sum()

0

---

* Inspect `GarageYrBlt` variable. Year garage was built

In [36]:
df['GarageYrBlt'].isna().sum()

81

In [37]:
df['GarageYrBlt'].value_counts().sort_index()

1900.0     1
1906.0     1
1908.0     1
1910.0     3
1914.0     2
          ..
2006.0    59
2007.0    49
2008.0    29
2009.0    21
2010.0     3
Name: GarageYrBlt, Length: 97, dtype: int64

In [38]:
df.loc[df.GarageYrBlt.isnull(),'GarageFinish'].value_counts()

None    81
Name: GarageFinish, dtype: int64

In [39]:
df[df.GarageFinish=='None']

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
39        1152       0.0           0.0         None           0         None   
48         736     716.0           2.0           No           0          Unf   
78        1768       0.0           4.0           No           0          Unf   
88        1013       0.0           2.0           No           0          Unf   
89         990       0.0           3.0           No         588          NaN   
...        ...       ...           ...          ...         ...          ...   
1349       938    1215.0           4.0           No         259          LwQ   
1407       833       0.0           3.0           No         625          NaN   
1449       630       0.0           1.0           Av         553          GLQ   
1450       896     896.0           4.0           No           0          Unf   
1453      1140       0.0           3.0           No           0          Unf   

      BsmtUnfSF  GarageArea GarageFinish  GarageYrBlt  ...  LotArea  \
39            0           0         None          NaN  ...     6040   
48          736           0         None          NaN  ...     4456   
78         1768           0         None          NaN  ...    10778   
88         1013           0         None          NaN  ...     8470   
89          402           0         None          NaN  ...     8070   
...         ...         ...          ...          ...  ...      ...   
1349        425           0         None          NaN  ...     5250   
1407        208           0         None          NaN  ...     8780   
1449         77           0         None          NaN  ...     1533   
1450        896           0         None          NaN  ...     9000   
1453       1140           0         None          NaN  ...    17217   

     LotFrontage  MasVnrArea  OpenPorchSF  OverallCond  OverallQual  \
39          65.0         0.0            0            5            4   
48          33.0         0.0            0            5            4   
78          72.0         0.0            0            5            4   
88         105.0         0.0            0            2            3   
89          60.0         0.0            0            5            4   
...          ...         ...          ...          ...          ...   
1349        50.0         0.0           54            5            8   
1407         NaN         0.0            0            5            5   
1449        21.0         0.0            0            7            5   
1450        60.0         0.0           45            5            5   
1453        90.0         0.0           56            5            5   

      TotalBsmtSF  YearBuilt  YearRemodAdd  SalePrice  
39              0       1955          1955      82000  
48            736       1920          2008     113000  
78           1768       1968          1968     136500  
88           1013       1915          1982      85000  
89            990       1994          1995     123600  
...           ...        ...           ...        ...  
1349          684       1872          1987     122000  
1407          833       1985          1985     112000  
1449          630       1970          1970      92000  
1450          896       1974          1974     136000  
1453         1140       2006          2006      84500  

[81 rows x 22 columns]

* We note there `81` null records for the `GarageYrBlt` and where the this variable is null `GarageFinish` is `None` meaning there is no garage.
* At this time there is nothing that can be done to enhance this field's analytical value 

---

# Push files to Repo

* In case you don't need to push files to Repo, you may replace this section for "Conclusions and Next Steps" and state your conclusions and next steps.

In [40]:
import os
try:
  os.makedirs(name='outputs/datasets') # create outputs/datasets/collection folder
except Exception as e:
  print(e)

* output the clean datasets to csv files into the outputs/datasets folder
* ./outputs/datasets/clean_house_prices_records.csv
* ./outputs/datasets/clean_inherited_houses.csv

In [41]:
df.to_csv("outputs/datasets/clean_house_prices_records.csv", index=False)

In [42]:
df_inherited.to_csv("outputs/datasets/clean_inherited_houses.csv", index=False)

---

# Conclusions and Next Steps


* Now we move on to the sales price study

---